### ISSUE TITLE
Allow resolvers to easily see what child fields are required

### ISSUE THREAD

**Issue by timothycoleman** [graphql-dotnet/graphql-dotnet#354]
In #21 there was some discussion around having a `context.GetChildFields()` method. #21 has been closed and broken up into subtickets but I don't think there is one for this.

In its `ResolveFieldAsync` method, the `DocumentExecutor` receives an argument `fields` which are the fields that the executor will recurse into. Perhaps it could put those fields into the `ResolveFieldContext` that it creates, so that the resolvers can determine (without having to traverse the AST again) whether particular child fields will be required.

**Comment by Jonatthu** [comment]
I really would like to have this feature

**Comment by Jonatthu** [comment]
@joemcbride How do we access to the subchilds currently?

**Comment by joemcbride** [comment]
@Jonatthu You don't.  The framework has access to them, though the field resolvers currently do not.

**Comment by Jonatthu** [comment]
@joemcbride If we add this, we could ad include parameters to our services to do joins in the same query, like one-to-one relationships this would be perfect, and for lists is other history...

**Comment by joemcbride** [comment]
Adding a note here: I was partially wrong in my previous statement that you don't have access to sub selections.  You DO have access to `FieldAst.SelectionSet`, which as the selections.  Those selections are not filtered by directives nor are fragments expanded, but you can get at the raw selections there.

**Comment by joemcbride** [comment]
Also, a limitation of having this feature (getting a list of child fields) is that it won't work with fields that return abstract types (interfaces and unions).  This is because the framework doesn't know what `GraphType` to use for the child selections until the parent resolver has returned a value.  Could perhaps provide a `ResolveFieldContext.GetChildFields(object source)` which took an object so that could be determined, though that may undermine the point of the feature (fetch child fields at the same time as the parent?)

**Comment by vertonghenb** [comment]
I've been looking into this situation and checking if we could use it to optimize our queries.
What I'm trying to do is getting the Children of the SelectionSet to create a Include string (IQueryable situation for EF).
Let's say we have the following query:
`
  	order(id:1) {
  		orderlines{
  			amount
  			product{
  				id
  				name
  			}
  		}
  		customer{
  			name
  			vatNo
  		}
  	}
`
In an ideal Linq to SQL situation we'd do the following:
dbContext.Orders.Include("orderlines.product,customer").Where(order => order.id == 1).ToList();
(note the '.' between Orderlines and Product

In other words if we can get the Navigation Properties of the query into the Include string, we can optimize quite a bunch of queries thus eliminating the N+1 in some situations.

I'm thinking to use reflection to get the navigation properties, but it would be sweet to get a parent-child relationship from the GraphQL selectionset to match it.

@Jonatthu  would something like this help you?
@joemcbride what do you think?

**Comment by flieks** [comment]
also interested, any progress ?
Now if we request 1 field of a child object, the resolver is getting the complete object first. 
Would be awesome if graphql-dotnet resolvers can return iqueryables and only fetches what it needs.

**Comment by joemcbride** [comment]
2.0 now provides a SubFields property in the resolver context.  There was also a major restructuring to support the DataLoader pattern, which also supplies parent node information.

See the docs on DataLoader: https://graphql-dotnet.github.io/dataloader/

And this tutorial: http://fiyazhasan.me/tag/graphql-dotnet/

**Comment by sungam3r** [comment]
Answered in https://github.com/graphql-dotnet/graphql-dotnet/issues/354#issuecomment-404857151 .
